File: android/src/main/java/com/bitcoinerlab/bitboxreactnative/BitcoinerlabBitBoxModule.kt
package com.bitcoinerlab.bitboxreactnative

import android.Manifest
import android.os.Build
import com.bitcoinerlab.bitboxreactnative.go.bitboxnative.Bitboxnative
import com.bitcoinerlab.bitboxreactnative.go.bitboxnative.Client
import com.bitcoinerlab.bitboxreactnative.go.bitboxnative.MobileTransport
import expo.modules.interfaces.permissions.PermissionsResponse
import expo.modules.interfaces.permissions.PermissionsResponseListener
import expo.modules.interfaces.permissions.PermissionsStatus
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.io.File
import org.json.JSONArray
import org.json.JSONObject

private const val BITBOX_NATIVE_MODULE_NAME = "BitcoinerlabBitBox"
private const val BITBOX_DEFAULT_CONNECT_TIMEOUT_MS = 30_000

class BitBoxNativeException(message: String, cause: Throwable? = null) :
  CodedException(message, cause)

data class BitBoxProductInfo(val product: String, val version: String)

private data class BitBoxSession(
  val id: String,
  val transportName: String,
  val transport: MobileTransport,
  val client: Client,
  val product: String,
  val version: String
)

private class BitBoxSessionStore {
  private val sessions = ConcurrentHashMap<String, BitBoxSession>()

  fun insert(
    transportName: String,
    transport: MobileTransport,
    client: Client,
    product: String,
    version: String
  ): BitBoxSession {
    val session = BitBoxSession(
      id = UUID.randomUUID().toString(),
      transportName = transportName,
      transport = transport,
      client = client,
      product = product,
      version = version
    )
    sessions[session.id] = session
    return session
  }

  fun session(sessionId: String): BitBoxSession = sessions[sessionId]
    ?: throw BitBoxNativeException("Unknown BitBox session: $sessionId")

  fun remove(sessionId: String) {
    val session = sessions.remove(sessionId)
      ?: throw BitBoxNativeException("Unknown BitBox session: $sessionId")
    session.client.close()
    session.transport.close()
  }
}

private fun bitboxTimeoutMs(params: Map<String, Any?>): Int {
  val timeoutMs = params["timeoutMs"]
  return when (timeoutMs) {
    is Number -> timeoutMs.toInt()
    else -> BITBOX_DEFAULT_CONNECT_TIMEOUT_MS
  }.coerceAtLeast(1)
}

private fun bitboxBytes(bytes: List<*>): ByteArray = bytes.map { value ->
  val byteValue = when (value) {
    is Number -> value.toInt()
    else -> throw BitBoxNativeException("BitBox byte arrays must contain numbers")
  }
  if (byteValue < 0 || byteValue > 255) {
    throw BitBoxNativeException("BitBox byte arrays must contain values from 0 to 255")
  }
  byteValue.toByte()
}.toByteArray()

// JS sends connect params as JSON so the React Native bridge only receives a
// string, never an object with nested undefined values.
private fun bitboxConnectParams(paramsJSON: String): Map<String, Any?> =
  if (paramsJSON.isEmpty()) emptyMap() else bitboxJSONObject(paramsJSON)

private fun bitboxJSONObject(string: String): Map<String, Any?> =
  bitboxJSONObjectToMap(JSONObject(string))

private fun bitboxJSONObjectToMap(jsonObject: JSONObject): Map<String, Any?> {
  val result = mutableMapOf<String, Any?>()
  val keys = jsonObject.keys()
  while (keys.hasNext()) {
    val key = keys.next()
    result[key] = bitboxJSONValue(jsonObject.get(key))
  }
  return result
}

private fun bitboxJSONArrayToList(jsonArray: JSONArray): List<Any?> =
  (0 until jsonArray.length()).map { index -> bitboxJSONValue(jsonArray.get(index)) }

private fun bitboxJSONValue(value: Any?): Any? = when (value) {
  null, JSONObject.NULL -> null
  is JSONObject -> bitboxJSONObjectToMap(value)
  is JSONArray -> bitboxJSONArrayToList(value)
  else -> value
}

private fun requiredBlePermissions(): Array<String> = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
  arrayOf(Manifest.permission.BLUETOOTH_SCAN, Manifest.permission.BLUETOOTH_CONNECT)
} else {
  arrayOf(Manifest.permission.ACCESS_FINE_LOCATION)
}

private fun ensureBlePermissions(module: Module, timeoutMs: Int) {
  val permissionsManager = module.appContext.permissions
    ?: throw BitBoxNativeException("Expo permissions service is not available; cannot request Android Bluetooth permissions")
  val permissions = requiredBlePermissions()
  val missingManifestPermissions = permissions.filterNot { permission ->
    permissionsManager.isPermissionPresentInManifest(permission)
  }
  if (missingManifestPermissions.isNotEmpty()) {
    throw BitBoxNativeException(
      "Missing Android Bluetooth manifest permissions: ${missingManifestPermissions.joinToString()}. Add @bitcoinerlab/bitbox-react-native/app.plugin to the Expo config."
    )
  }
  if (permissionsManager.hasGrantedPermissions(*permissions)) return

  val latch = CountDownLatch(1)
  var result: Map<String, PermissionsResponse>? = null
  permissionsManager.askForPermissions(
    PermissionsResponseListener { response ->
      result = response
      latch.countDown()
    },
    *permissions
  )
  if (!latch.await(timeoutMs.toLong(), TimeUnit.MILLISECONDS)) {
    throw BitBoxNativeException("Timed out waiting for Android Bluetooth permission response")
  }
  val denied = permissions.filter { permission ->
    result?.get(permission)?.status != PermissionsStatus.GRANTED
  }
  if (denied.isNotEmpty()) {
    throw BitBoxNativeException("Android Bluetooth permissions were denied: ${denied.joinToString()}")
  }
}

class BitcoinerlabBitBoxModule : Module() {
  private val sessions = BitBoxSessionStore()

  override fun definition() = ModuleDefinition {
    Name(BITBOX_NATIVE_MODULE_NAME)

    AsyncFunction("connectBle") { paramsJSON: String ->
      val params = bitboxConnectParams(paramsJSON)
      val timeoutMs = bitboxTimeoutMs(params)
      ensureBlePermissions(this@BitcoinerlabBitBoxModule, timeoutMs)
      val context = appContext.reactContext
        ?: throw BitBoxNativeException("React context is not available")
      val transport = BitBoxBleTransport(
        context = context,
        timeoutMs = timeoutMs,
        deviceId = params["deviceId"] as? String
      )
      val productInfo = try {
        transport.connect()
      } catch (error: Throwable) {
        try { transport.close() } catch (_: Throwable) {}
        throw error
      }
      val client = try {
        Bitboxnative.newClientWithMobileTransport(
          transport,
          productInfo.product,
          productInfo.version,
          true
        )
      } catch (error: Throwable) {
        try { transport.close() } catch (_: Throwable) {}
        throw error
      }
      val session = sessions.insert(
        transportName = "ble",
        transport = transport,
        client = client,
        product = productInfo.product,
        version = productInfo.version
      )
      mapOf(
        "id" to session.id,
        "transport" to session.transportName,
        "product" to session.product,
        "version" to session.version
      )
    }

    AsyncFunction("connectUsb") { paramsJSON: String ->
      val params = bitboxConnectParams(paramsJSON)
      val timeoutMs = bitboxTimeoutMs(params)
      val activity = appContext.throwingActivity
      val context = appContext.reactContext
        ?: throw BitBoxNativeException("React context is not available")
      val transport = BitBoxUsbTransport(
        context = context,
        timeoutMs = timeoutMs,
        deviceId = params["deviceId"] as? String
      )
      val productInfo = try {
        transport.connect()
      } catch (error: Throwable) {
        try { transport.close() } catch (_: Throwable) {}
        throw error
      }
      val client = try {
        Bitboxnative.newClientWithMobileTransportAndPairingConfig(
          transport,
          "",
          "",
          false,
          BitBoxPairingConfirmation(activity, timeoutMs),
          File(context.filesDir, "bitbox-usb-noise-config.json").absolutePath
        )
      } catch (error: Throwable) {
        try { transport.close() } catch (_: Throwable) {}
        throw error
      }
      val version = client.version()
      val session = sessions.insert(
        transportName = "usb",
        transport = transport,
        client = client,
        product = productInfo.product,
        version = version
      )
      mapOf(
        "id" to session.id,
        "transport" to session.transportName,
        "product" to session.product,
        "version" to session.version
      )
    }

    AsyncFunction("disconnect") { sessionId: String ->
      sessions.remove(sessionId)
    }

    AsyncFunction("version") { sessionId: String ->
      sessions.session(sessionId).client.version()
    }

    AsyncFunction("rootFingerprint") { sessionId: String ->
      sessions.session(sessionId).client.rootFingerprint()
    }

    AsyncFunction("btcXpub") { sessionId: String,
      apiNetwork: String,
      keypath: String,
      xpubType: String,
      display: Boolean ->
      sessions.session(sessionId).client.btcxPub(
        apiNetwork,
        keypath,
        xpubType,
        display
      )
    }

    // BitBox script configs arrive as JSON made in JS. This keeps parameter
    // passing predictable on Android and matches the Go mobile adapter.
    AsyncFunction("btcAddress") { sessionId: String,
      apiNetwork: String,
      keypath: String,
      scriptConfigJSON: String,
      display: Boolean ->
      sessions.session(sessionId).client.btcAddress(
        apiNetwork,
        keypath,
        scriptConfigJSON,
        display
      )
    }

    AsyncFunction("btcRegisterScriptConfig") { sessionId: String,
      apiNetwork: String,
      scriptConfigJSON: String,
      keypathAccount: String,
      xpubType: String,
      name: String ->
      sessions.session(sessionId).client.btcRegisterScriptConfig(
        apiNetwork,
        scriptConfigJSON,
        keypathAccount,
        xpubType,
        name
      )
    }

    AsyncFunction("btcIsScriptConfigRegistered") { sessionId: String,
      apiNetwork: String,
      scriptConfigJSON: String,
      keypathAccount: String ->
      sessions.session(sessionId).client.btcIsScriptConfigRegistered(
        apiNetwork,
        scriptConfigJSON,
        keypathAccount
      )
    }

    AsyncFunction("btcSignPSBT") { sessionId: String,
      apiNetwork: String,
      psbt: String,
      forceScriptConfigJSON: String,
      formatUnit: String ->
      sessions.session(sessionId).client.btcSignPSBT(
        apiNetwork,
        psbt,
        forceScriptConfigJSON,
        formatUnit
      )
    }

    AsyncFunction("btcSignMessage") { sessionId: String,
      apiNetwork: String,
      scriptConfigWithKeypathJSON: String,
      message: List<Any> ->
      val resultJSON = sessions.session(sessionId).client.btcSignMessage(
        apiNetwork,
        scriptConfigWithKeypathJSON,
        bitboxBytes(message)
      )
      bitboxJSONObject(resultJSON)
    }
  }
}

File: ios/BitcoinerlabBitBoxModule.swift
import Bitboxnative
import ExpoModulesCore
import Foundation

private let bitboxNativeModuleName = "BitcoinerlabBitBox"
private let bitboxDefaultConnectTimeoutMs = 30_000
private let bitboxDefaultScanDurationMs = 5_000

private func bitboxPositiveInt(
  _ name: String,
  from params: [String: Any],
  defaultValue: Int
) -> Int {
  if let value = params[name] as? Int {
    return max(value, 1)
  }
  if let value = params[name] as? Double {
    return max(Int(value), 1)
  }
  if let value = params[name] as? NSNumber {
    return max(value.intValue, 1)
  }
  return defaultValue
}

private final class BitBoxSession {
  let id: String
  let transport: BitBoxBleTransport
  let client: BitcoinerlabBitBoxBitboxnativeClient
  let product: String
  let version: String

  init(
    id: String,
    transport: BitBoxBleTransport,
    client: BitcoinerlabBitBoxBitboxnativeClient,
    product: String,
    version: String
  ) {
    self.id = id
    self.transport = transport
    self.client = client
    self.product = product
    self.version = version
  }
}

private final class BitBoxSessionStore {
  private let lock = NSLock()
  private var sessions: [String: BitBoxSession] = [:]

  func insert(
    transport: BitBoxBleTransport,
    client: BitcoinerlabBitBoxBitboxnativeClient,
    product: String,
    version: String
  ) -> BitBoxSession {
    let session = BitBoxSession(
      id: UUID().uuidString,
      transport: transport,
      client: client,
      product: product,
      version: version
    )
    lock.lock()
    sessions[session.id] = session
    lock.unlock()
    return session
  }

  func session(_ sessionId: String) throws -> BitBoxSession {
    lock.lock()
    let session = sessions[sessionId]
    lock.unlock()
    guard let session = session else {
      throw BitBoxNativeError(message: "Unknown BitBox session: \(sessionId)")
    }
    return session
  }

  func remove(_ sessionId: String) throws {
    lock.lock()
    let session = sessions.removeValue(forKey: sessionId)
    lock.unlock()
    guard let session = session else {
      throw BitBoxNativeError(message: "Unknown BitBox session: \(sessionId)")
    }
    session.client.close()
    try? session.transport.close()
  }
}

private func bitboxCallString(_ call: (NSErrorPointer) -> String) throws -> String {
  var error: NSError?
  let result = call(&error)
  if let error = error {
    throw error
  }
  return result
}

private func bitboxBytes(_ bytes: [Int]) throws -> [UInt8] {
  try bytes.map { byte in
    guard byte >= 0, byte <= 255 else {
      throw BitBoxNativeError(message: "BitBox byte arrays must contain values from 0 to 255")
    }
    return UInt8(byte)
  }
}

// JS sends connect params as JSON so the React Native bridge only receives a
// string, never an object with nested undefined values.
private func bitboxConnectParams(from paramsJSON: String) throws -> [String: Any] {
  if paramsJSON.isEmpty {
    return [:]
  }
  return try bitboxJSONObject(paramsJSON)
}

private func bitboxJSONObject(_ string: String) throws -> [String: Any] {
  let data = Data(string.utf8)
  guard let object = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] else {
    throw BitBoxNativeError(message: "BitBox JSON result was not an object")
  }
  return object
}

public class BitcoinerlabBitBoxModule: Module {
  private let sessions = BitBoxSessionStore()

  public func definition() -> ModuleDefinition {
    Name(bitboxNativeModuleName)

    AsyncFunction("discoverBle") { (_ paramsJSON: String) throws -> [[String: Any]] in
      let params = try bitboxConnectParams(from: paramsJSON)
      return try BitBoxBleDiscovery(timeoutMs: bitboxDefaultConnectTimeoutMs).discover(
        scanDurationMs: bitboxPositiveInt(
          "scanDurationMs",
          from: params,
          defaultValue: bitboxDefaultScanDurationMs
        )
      )
    }

    AsyncFunction("listUsb") { () throws -> [[String: Any]] in
      throw BitBoxNativeError(message: "USB is not supported on iOS")
    }

    AsyncFunction("connectBle") { (_ paramsJSON: String) throws -> [String: Any] in
      let params = try bitboxConnectParams(from: paramsJSON)
      let transport = BitBoxBleTransport(
        timeoutMs: bitboxPositiveInt(
          "timeoutMs",
          from: params,
          defaultValue: bitboxDefaultConnectTimeoutMs
        )
      )
      let productInfo = try transport.connect(deviceId: params["deviceId"] as? String)
      var goError: NSError?
      guard
        let client = BitcoinerlabBitBoxBitboxnativeNewClientWithMobileTransport(
          transport,
          productInfo.product,
          productInfo.version,
          true,
          &goError
        )
      else {
        try? transport.close()
        throw goError ?? BitBoxNativeError(message: "Failed to initialize BitBox Nova BLE session")
      }

      let session = self.sessions.insert(
        transport: transport,
        client: client,
        product: productInfo.product,
        version: productInfo.version
      )
      return [
        "id": session.id,
        "transport": "ble",
        "product": session.product,
        "version": session.version
      ]
    }

    AsyncFunction("connectUsb") { (_ paramsJSON: String) throws -> [String: Any] in
      _ = paramsJSON
      throw BitBoxNativeError(message: "USB is not supported on iOS; use connectBitBoxNovaBle")
    }

    AsyncFunction("disconnect") { (_ sessionId: String) throws in
      try self.sessions.remove(sessionId)
    }

    AsyncFunction("version") { (_ sessionId: String) throws -> String in
      let client = try self.sessions.session(sessionId).client
      return try bitboxCallString { error in
        client.version(error)
      }
    }

    AsyncFunction("rootFingerprint") { (_ sessionId: String) throws -> String in
      let client = try self.sessions.session(sessionId).client
      return try bitboxCallString { error in
        client.rootFingerprint(error)
      }
    }

    AsyncFunction("btcXpub") { (
      _ sessionId: String,
      _ apiNetwork: String,
      _ keypath: String,
      _ xpubType: String,
      _ display: Bool
    ) throws -> String in
      let client = try self.sessions.session(sessionId).client
      return try bitboxCallString { error in
        client.btcxPub(
          apiNetwork,
          keypath: keypath,
          xpubType: xpubType,
          display: display,
          error: error
        )
      }
    }

    // BitBox script configs arrive as JSON made in JS. This keeps parameter
    // passing predictable on both native platforms and matches the Go adapter.
    AsyncFunction("btcAddress") { (
      _ sessionId: String,
      _ apiNetwork: String,
      _ keypath: String,
      _ scriptConfigJSON: String,
      _ display: Bool
    ) throws -> String in
      let client = try self.sessions.session(sessionId).client
      return try bitboxCallString { error in
        client.btcAddress(
          apiNetwork,
          keypath: keypath,
          scriptConfigJSON: scriptConfigJSON,
          display: display,
          error: error
        )
      }
    }

    AsyncFunction("btcRegisterScriptConfig") { (
      _ sessionId: String,
      _ apiNetwork: String,
      _ scriptConfigJSON: String,
      _ keypathAccount: String,
      _ xpubType: String,
      _ name: String
    ) throws in
      let client = try self.sessions.session(sessionId).client
      try client.btcRegisterScriptConfig(
        apiNetwork,
        scriptConfigJSON: scriptConfigJSON,
        keypathAccount: keypathAccount,
        xpubType: xpubType,
        name: name
      )
    }

    AsyncFunction("btcIsScriptConfigRegistered") { (
      _ sessionId: String,
      _ apiNetwork: String,
      _ scriptConfigJSON: String,
      _ keypathAccount: String
    ) throws -> Bool in
      let client = try self.sessions.session(sessionId).client
      var registered = ObjCBool(false)
      try client.btcIsScriptConfigRegistered(
        apiNetwork,
        scriptConfigJSON: scriptConfigJSON,
        keypathAccount: keypathAccount,
        ret0_: &registered
      )
      return registered.boolValue
    }

    AsyncFunction("btcSignPSBT") { (
      _ sessionId: String,
      _ apiNetwork: String,
      _ psbt: String,
      _ forceScriptConfigJSON: String,
      _ formatUnit: String
    ) throws -> String in
      let client = try self.sessions.session(sessionId).client
      return try bitboxCallString { error in
        client.btcSignPSBT(
          apiNetwork,
          psbtBase64: psbt,
          forceScriptConfigJSON: forceScriptConfigJSON,
          formatUnit: formatUnit,
          error: error
        )
      }
    }

    AsyncFunction("btcSignMessage") { (
      _ sessionId: String,
      _ apiNetwork: String,
      _ scriptConfigWithKeypathJSON: String,
      _ message: [Int]
    ) throws -> [String: Any] in
      let client = try self.sessions.session(sessionId).client
      let messageData = Data(try bitboxBytes(message))
      let resultJSON = try bitboxCallString { error in
        client.btcSignMessage(
          apiNetwork,
          scriptConfigWithKeypathJSON: scriptConfigWithKeypathJSON,
          message: messageData,
          error: error
        )
      }
      return try bitboxJSONObject(resultJSON)
    }
  }
}
